README.md

CPUTune
====
An open source kernel extension enables dynamic CPU performance tuning at runtime for macOS.

#### Features
- Allows turning on or off Intel Turbo Boost for better battery life
- Allows turning on or off Intel Speed Shift for maximum performance
- Provides the fine granularity to change HWP (performance value) at runtime
- Provides the control over tuning turbo ratio limit at runtime when turbo boost is enabled
- Allows turning on or off Intel Proc Hot (```BD_PROCHOT```) for running without a battery. Please note that this is NOT recommended and can seriously damage your mac. Do at your own risk.
- Implements TimerEvent-based responses for dynamical switching Turbo Boost and Speed Shift at runtime
- Allows System Integrity Protection (SIP) control a bit easier via Info.plist setting 

#### Boot arguments
- Add `-cputoff` to disable CPUTune
- Add `-cputbeta` to enable CPUTune on unsupported os versions (10.15 and below are enabled by default).

#### Configuration
Open terminal:
- Type in ```echo 1>/tmp/CPUTuneTurboBoostRT.conf``` to enable turbo boost when CPUTune is loaded
- Type in ```echo 0>/tmp/CPUTuneTurboBoostRT.conf``` to disable turbo boost when needed
- Type in ```echo <request value> >/tmp/HWPRequest.conf``` to submit persistency hwp request at runtime. For example ```<requet value> = 0x80193008```
- Type in ```echo <turbo ratio limit> >/tmp/TurboRatioLimit.conf``` to submit cores maximum frequency at runtime. For example ```0x2b2c2d2f3030``` in ```i9-8950HK (6 cores)``` sets the maximum frequency for Core[1-6] at 43(0x2b), 44(0x2c), 45(0x2d), 46(0x2f), 48(0x30), 48(0x30) respectively.
- Type in  ```echo 1>/tmp/CPUTuneProcHotRT.conf``` to enable proc hot when needed
- Type in  ```echo 0>/tmp/CPUTuneProcHotRT.conf``` to siable proc hot when needed
- Change update time interval (millisecond) in `CPUTune.kext/Contents/Info.plist` to have a more  looser/tigher control over HWP request
- In case you want a simplify command to switch turbo boost, change the `TurboBoostAtRuntime` in `CPUTune.kext/Contents/Info.plist`

#### Contribution
All suggestions and improvements are welcome, don't hesitate to pull request or open an issue if you want this project to be better than ever.
Writing and supporting code is fun but it takes time. Please provide most descriptive bugreports or pull requests.


#### Change Log
[Change logs](https://github.com/syscl/CPUTune/blob/master/Changelog.md) for detail improvements



Changelog.md
CPUTune Changelog
=======================
#### v 2.2.6

- Corrected the timeout unit from nano to second to fix the memory leak

#### v2.2.5

- Added missing semicolon

#### v2.2.4

- Fixed the type casting issue for buffer in readFileAsBytes()

#### v2.2.3

- Attempt to fix the memory leak in readFileAsBytes()

#### v2.2.2

- Renamed CPUTune.kext to CPUTuneCore.kext

#### v2.2.1

- Fixed a kernel panic when resart the timer in readConfigAtRuntime()

#### v2.2.0

- Fixed return status in start()
- Reverted to less aggressive MSR writters with null pointer protection
- Removed redudant log in setIfNotEqual

#### v2.1.9

- Fixed some nits

#### v2.1.8

- Protected against null pointer
- Fixed the readConfigAtRuntime read and set logic (use setIfNotEqual)
- Added Base to know region as new Xcode

#### v 2.1.7

- Fixed the kernel panic at boostrap due to write log to filesystem

#### v 2.1.6

- Fixed a kernel panic that reads MSR_IA32_HWP_REQUEST on CPUs unsupport HWP

#### v2.1.5

- Fixed the NVRAM log string buffer

#### v2.1.4

- Add version info for CPUTune; rename readFileNBytes to readFileAsBytes 

#### v2.1.3

- Add line number, better code style for LOG and DBGLOG

#### v2.1.2

- Better logging with function name, more readable

#### v2.1.1

- Fixed timer sounce timeout in ```CPUTune::start```

#### v2.1.0

- Introduce ```DBGLOG()``` macro to reduce the log noise
- Rename ```myLOG()``` to ```LOG()```

#### v2.0.9

- Remove unused header in CPUInfo's header; bump to the right version

#### v2.0.8

- Check turbo rate limit RW/RO before writting to MSR

#### v2.0.7

- Fix memory leak by freeing buffer after use

#### v2.0.6

- Tune turbo ratio limit at runtime when turbo is enabled

#### v2.0.5

- Make code cleaner and shorter, remove unnecessary ```initKextPreferences()```

#### v2.0.4

- Make update timeout configurable so that we have a more looser/tigher control over the HWP request
- Move the ```initKextPreferences()``` up to ```CPUTune::init()```

#### v2.0.3

- Decouple HWP information to ```CPUInfo```

#### v2.0.2

- Support no ```0x/0X``` prefix hexadecimal constant (e.g. 0x800d3008 as 800d3008)

#### v2.0.1

- Verify and surface invalid HWP request properly

#### v2.0.0

- Major release, fix logging in current value after ```setIfNotEqual()```

#### v1.9.9

- Introduce HWP request at runtime

#### v1.9.8

- Avoid construct objects twice

#### v1.9.7

- Remove useless const qualifier + improved code style

#### v1.9.6

- Added init/restore for ```BD_PROCHOT```
- Unified the log output for PROCHOT and use ```setIfNotEqual()```
- Fixed a bug in ```readConfigAtRuntime``` which was not restoring ProcHot correctly when coming out of sleep mode

#### v1.9.5

- Fixed the return type of the ```setIfNotEqual()```

#### v1.9.4

- Simplify writing logic to MSR

#### v1.9.3
- Avoid dynamic allocation for CPUInfo and SIPTune, which avoid indefine blocking in kernel space + avoid memory leak
- Use type anotation

#### v1.9.2

- More accurate logging for non-exist propreties in NVRAM

#### v1.9.1

- Supports ```BD_PROCHOT``` signal on Intel CPU credits @christophe-duc
- Update README.md

#### v1.9.0

- Resolve linkage issue for kernel major/minor

#### v1.8.9

- NVRAM protection

#### v1.8.8

- Supported macOS Catalina

#### v1.8.7

- Updated cpu info to ```xnu-4903.221.2```

#### v1.8.6

- Updated csr header to ```xnu-4903.221.2```

#### v.1.8.5

- Simplify the enable if-condition

#### v1.8.4

- Removed unused variables
- Corrected  ```writeBufferToFile()``` return type

#### v1.8.1
- Regularly checks profiles

#### v1.8.0
- Allowed SIP control a bit easier via Info.plist tunning instead of reboot to recovery mode on a real mac
- Simplify memory deallocate code blocks via deleter routine

#### v1.7.0
- Fixed compatible issues (supported OS X Mountain Lion and Xcode 3.2)
- Fixed code logic of HWP Enable, details can be found [here](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3b-part-2-manual.pdf)

#### v1.6.5
- Implemented Timer Event based feature for dynamic control CPU Performance at runtime 
- CPUInfo to detect cpu information

#### v1.1.1
- Implemented write logs to file

#### v1.0.7
- Boot arguments check 
- Unsupported version check 
- Fixed memeory leak

#### v1.0.1
- Fixed enableTurboBoost().

#### v1.0.0
- Initial release


CPUTune/CPUInfo.hpp
//
//  CPUInfo.hpp
//  CPUTune
//
//  Copyright (c) 2018 syscl. All rights reserved.
//

#ifndef CPUInfo_hpp
#define CPUInfo_hpp

#include "kern_util.hpp"
#include <i386/cpuid.h>
/* Copied from xnu/osfmk/cpuid.c (modified for 64-bit values) */
#define bit(n)                 (1UL << (n))
#define bitmask64(h, l)        ((bit(h) | (bit(h) - 1)) & ~ (bit(l) - 1))
#define bitfield32(x, h, l)    (((x) & bitmask64(h, l)) >> l)

// As per xnu/osfmk/proc_reg.h
#define MSR_CORE_THREAD_COUNT       0x35

// Intel SpeedShift MSRs
#define MSR_IA32_PM_ENABLE          0x770
#define MSR_IA32_HWP_REQUEST        0x774

// Intel Power MSRs
#define MSR_IA32_POWER_CTL          0x1FC

// Maximum Ratio Limit of Turbo Mode
// RW or RO stores in MSR_PLATFORM_INFO.[28]
// Refer Software Developer's Manual Volume 4: Model-Specific Registers
#define MSR_TURBO_RATIO_LIMIT_RW    (1 << 28)
#define MSR_TURBO_RATIO_LIMIT       0x1AD
#define MSR_TURBO_RATIO_LIMIT1      0x1AE
#define MSR_TURBO_RATIO_LIMIT2      0x1AF

class CPUInfo {
public:
    CPUInfo() :
        model(getCPUModel()),
        supportedHWP(supportedSpeedShift()),
        coreCount(getCoreCount()),
        turboRatioLimitRW(getTurboRatioLimitRW()) {
        LOG("cpu model: 0x%x, %s HWP, number of cores: %d, turbo ratio limit permission: %s",
              model,
              (supportedHWP ? "supported" : "unsupported"),
              coreCount,
              (turboRatioLimitRW ? "RW" : "RO"));
    };
    
    /**
     * Intel CPU models
     */
    const uint8_t model;
    
    /**
     *  CPU support HWP
     */
    const bool supportedHWP;
    
    /**
     * CPU Cores Count
     */
    const uint8_t coreCount;
    
    const bool turboRatioLimitRW;
    
    /**
     *  Get current CPU model.
     *
     *  @return detected CPU model
     */
    const uint8_t getCPUModel(void) const;
    
    const bool supportedSpeedShift(void) const;
    
    const uint8_t getCoreCount(void) const;
    
    const bool getTurboRatioLimitRW(void) const;
    
    /**
    *  Intel CPU models as returned by CPUID
    *  The list is synchronised and updated with XNU source code (osfmk/i386/cpuid.h).
    *  Names are altered to avoid conflicts just in case.
    *  Last update: xnu-4903.221.2
    *  Some details could be found on http://instlatx64.atw.hu and https://en.wikichip.org/wiki/64-bit_architecture#x86
    *  Also: https://www.intel.com/content/dam/www/public/us/en/documents/sa00115-microcode-update-guidance.pdf
    */
    enum CpuModel {
        CPU_MODEL_UNKNOWN        =  0x00,
        CPU_MODEL_PENRYN         =  0x17,
        CPU_MODEL_NEHALEM        =  0x1A,
        CPU_MODEL_FIELDS         =  0x1E, /* Lynnfield, Clarksfield */
        CPU_MODEL_DALES          =  0x1F, /* Havendale, Auburndale */
        CPU_MODEL_NEHALEM_EX     =  0x2E,
        CPU_MODEL_DALES_32NM     =  0x25, /* Clarkdale, Arrandale */
        CPU_MODEL_WESTMERE       =  0x2C, /* Gulftown, Westmere-EP/-WS */
        CPU_MODEL_WESTMERE_EX    =  0x2F,
        CPU_MODEL_SANDYBRIDGE    =  0x2A,
        CPU_MODEL_JAKETOWN       =  0x2D,
        CPU_MODEL_IVYBRIDGE      =  0x3A,
        CPU_MODEL_IVYBRIDGE_EP   =  0x3E,
        CPU_MODEL_CRYSTALWELL    =  0x46,
        CPU_MODEL_HASWELL        =  0x3C,
        CPU_MODEL_HASWELL_EP     =  0x3F,
        CPU_MODEL_HASWELL_ULT    =  0x45,
        CPU_MODEL_BROADWELL      =  0x3D,
        CPU_MODEL_BROADWELL_ULX  =  0x3D,
        CPU_MODEL_BROADWELL_ULT  =  0x3D,
        CPU_MODEL_BRYSTALWELL    =  0x47,
        CPU_MODEL_SKYLAKE        =  0x4E,
        CPU_MODEL_SKYLAKE_ULT    =  0x4E,
        CPU_MODEL_SKYLAKE_ULX    =  0x4E,
        CPU_MODEL_SKYLAKE_DT     =  0x5E,
        CPU_MODEL_SKYLAKE_W      =  0x55,
        CPU_MODEL_KABYLAKE       =  0x8E,
        CPU_MODEL_KABYLAKE_ULT   =  0x8E,
        CPU_MODEL_KABYLAKE_ULX   =  0x8E,
        CPU_MODEL_KABYLAKE_DT    =  0x9E,
        CPU_MODEL_COMETLAKE_S    =  0x9F, /* desktop Comet Lake */
        CPU_MODEL_CANNONLAKE     =  0x66,
        CPU_MODEL_ICELAKE_Y      =  0x7D,
        CPU_MODEL_ICELAKE_U      =  0x7E,
        CPU_MODEL_COMETLAKE_Y    =  0xA5, /* aka 10th generation Amber Lake Y */
        CPU_MODEL_COMETLAKE_U    =  0xA6,
    };
};

#endif /* CPUInfo_hpp */
